cloud/wfm/src/process-instance/index.ts

```diff
@@ -0,0 +1,43 @@
+import * as Promise from 'bluebird';
+import {EventEmitter} from 'eventemitter3';
+import * as _ from 'lodash';
+import {Process} from '../process';
+import {Task} from '../task';
+
+export interface InstanceTaskEventData<T extends ProcessInstance>  {
+  instance: T;
+  task: Task;
+}
+
+export interface InstanceEventData<T extends ProcessInstance> {
+  instance: T;
+}
+
+/**
+ * The executable instance of a {@link Process}
+ */
+export interface ProcessInstance extends EventEmitter {
+  assigneeId: string;
+
+  getTasks(): Promise<Task[]>;
+  getCurrentTask(): Promise<Task>;
+  /**
+   * @return A promise that resolves to the next task in it's list,
+   * so implementation be async
+   */
+  next(): Promise<Task>;
+
+  getId(): string;
+
+  /**
+   * Event emitted when the current active task changes to the next one
+   */
+  on(event: 'taskChange', handler: (e: InstanceTaskEventData<this>) => any): this;
+  /**
+   * Event emitted when all {@link Task}s achieve the done state
+   */
+  on(event: 'done', handler: (e: InstanceEventData<this>) => any): this;
```

Comment:
I took a better look into this, and it's also a -1 from me. TS is actually quite good for documenting events. `instance.on('something-else')` is a compile error

Intellisense support is also good:

![image](https://user-images.githubusercontent.com/1294893/27353556-cb63bab6-55da-11e7-9fa1-ddc9474765bc.png)

Bear in mind that java uses listener interfaces because of language restrictions (functions aren't first class citizens, so they need interfaces + anonymous classes) and not because of OOP practices